crates/uv/src/commands/pip/tree.rs

```diff
@@ -0,0 +1,210 @@
+use std::fmt::Write;
+
+use anyhow::Result;
+use distribution_types::{Diagnostic, InstalledDist, Name};
+use owo_colors::OwoColorize;
+use tracing::debug;
+use uv_cache::Cache;
+use uv_configuration::PreviewMode;
+use uv_fs::Simplified;
+use uv_installer::SitePackages;
+use uv_normalize::PackageName;
+use uv_toolchain::Toolchain;
+use uv_toolchain::ToolchainRequest;
+use uv_toolchain::{PythonEnvironment, SystemPython};
+
+use crate::commands::ExitStatus;
+use crate::printer::Printer;
+use std::collections::{HashMap, HashSet};
+
+use pypi_types::VerbatimParsedUrl;
+
+/// Display the installed packages in the current environment as a dependency tree.
+pub(crate) fn pip_tree(
+    strict: bool,
+    python: Option<&str>,
+    system: bool,
+    preview: PreviewMode,
+    cache: &Cache,
+    printer: Printer,
+) -> anyhow::Result<ExitStatus> {
+    // Detect the current Python interpreter.
+    let system = if system {
+        SystemPython::Required
+    } else {
+        SystemPython::Allowed
+    };
+    let environment = PythonEnvironment::from_toolchain(Toolchain::find(
+        python.map(ToolchainRequest::parse),
+        system,
+        preview,
+        cache,
+    )?);
+
+    debug!(
+        "Using Python {} environment at {}",
+        environment.interpreter().python_version(),
+        environment.python_executable().user_display().cyan()
+    );
+
+    // Build the installed index.
+    let site_packages = SitePackages::from_executable(&environment)?;
+
+    match DisplayDependencyGraph::new(&site_packages).render() {
+        Ok(lines) => {
+            writeln!(printer.stdout(), "{}", lines.join("\n"))?;
+        }
+        Err(e) => {
+            writeln!(
+                printer.stderr(),
+                "Unable to display the dependency tree; {e}",
+            )
+            .unwrap();
+            return Ok(ExitStatus::Failure);
+        }
+    }
+
+    // Validate that the environment is consistent.
+    if strict {
+        for diagnostic in site_packages.diagnostics()? {
+            writeln!(
+                printer.stderr(),
+                "{}{} {}",
+                "warning".yellow().bold(),
+                ":".bold(),
+                diagnostic.message().bold()
+            )?;
+        }
+    }
+    Ok(ExitStatus::Success)
+}
+
+// Filter out all required packages of the given distribution if they
+// are required by an extra.
+// For example, `requests==2.32.3` requires `charset-normalizer`, `idna`, `urllib`, and `certifi` at
+// all times, `PySocks` on `socks` extra and `chardet` on `use_chardet_on_py3` extra.
+// This function will return `["charset-normalizer", "idna", "urllib", "certifi"]` for `requests`.
+fn required_with_no_extra(dist: &InstalledDist) -> Vec<pep508_rs::Requirement<VerbatimParsedUrl>> {
```

Comment:
What's the reason for filtering these out?

Reply:
cycles become way too common if extras are included (in which case the current implementation throws an error). `pipdeptree` seems to just remove the extras when building the tree, and it seemed like the right choice for usability.

Reply:
I see, that seems reasonable. But why filter only the extras that are included by default, and not all extras installed?

Reply:
hmm the code looks like filtering all extra packages altogether, but maybe i'm missing something?